source/ceylon/collection/ReversibleHashMap.ceylon

```diff
@@ -0,0 +1,110 @@
+"A [[Map]] which may be [[reversed|reverse]], backed by two
+ [[HashMap]]s. The reversible mapping from [[Key]]s to 
+ [[Item]]s is a _bijection_ (a one-to-one mapping).
+ 
+ In order to maintain the bijection when the 
+ `ReversibleHashMap` is mutated, the [[put]] operation not 
+ only overwrites any existing mapping for the given key, but
+ also removes any existing entry with the given item. Note
+ that this behavior, in principle, violates the general
+ contract of [[MutableMap.put]]."
+//TODO: generalize this implementation to also allow
+//      ReversibleTreeMaps!
+shared class ReversibleHashMap<Key, Item>
+        (stability = linked, hashtable = Hashtable(),
+         entries = {}) 
+        satisfies MutableMap<Key, Item> 
+        given Key satisfies Object
+        given Item satisfies Object {
+    
+    "Performance-related settings for the backing arrays."
+    Hashtable hashtable;
+    
+    "Determines whether this is a linked hash map with a
+     stable iteration order."
+    Stability stability;
+    
+    "The initial entries in the map."
+    {<Key->Item>*} entries;
+    
+    variable MutableMap<Key, Item> map 
+            = HashMap<Key, Item>
+                    (stability, hashtable, entries);
+    variable MutableMap<Item, Key> reverseMap 
+            = HashMap<Item, Key>
+                    (stability, hashtable);
+    for (key->item in entries) {
+        reverseMap.put(item, key);
+    }
+    
+    "The reverse mapping from [[Item]]s to [[Key]]s. This
+     operation is lazy, returning a view of this reversible
+     map."
+    shared ReversibleHashMap<Item,Key> reverse {
+        value reverse = ReversibleHashMap<Item,Key>
+                (stability, hashtable);
+        reverse.reverseMap = map;
+        reverse.map = reverseMap;
+        return reverse;
+    }
+    
+    size => map.size;
+    
+    defines(Object key) => map.defines(key);
+    
+    contains(Object entry) => map.contains(entry);
+    
+    get(Object key) => map.get(key);
+    
+    iterator() => map.iterator();
+    
+    keys => map.keys;
+    
+    items => reverseMap.keys;
+    
```

Comment:
Refine to `Category<Item>`? Or even `Set<Item>` for this and `keys` with additional support from `HashMap`.